chef/cookbooks/nova/recipes/compute.rb

```diff
@@ -106,167 +106,170 @@ def set_boot_kernel_and_trigger_reboot(flavor='default')
-            package pkg do
-              action :install
-            end
-          end
+    case node[:nova][:libvirt_type]
+    when "kvm", "qemu"
+      if node.platform_version.to_f >= 12.0
+        package "qemu"
+      else
+        package "kvm"
+      end
 
-          service "xend" do
-            action :nothing
-            supports :status => true, :start => true, :stop => true, :restart => true
-            # restart xend only when xen kernel is already present
-            only_if { %x[uname -r].include?('xen') }
-          end
+      set_boot_kernel_and_trigger_reboot
 
-          template "/etc/xen/xend-config.sxp" do
-            source "xend-config.sxp.erb"
-            group "root"
-            owner "root"
-            mode 0644
-            variables(
-              :node_platform => node[:platform],
-              :libvirt_migration => node[:nova]["use_migration"],
-              :shared_instances => node[:nova]["use_shared_instance_storage"],
-              :libvirtd_listen_tcp => node[:nova]["use_migration"] ? 1 : 0,
-              :libvirtd_listen_addr => Chef::Recipe::Barclamp::Inventory.get_network_by_type(node, "admin").address
-            )
-            notifies :restart, "service[xend]", :delayed
-          end
+      if node[:nova][:libvirt_type] == "kvm"
+        if node.platform_version.to_f >= 12.0
+          package "qemu-kvm" if node[:architecture] == "x86_64"
+          package "qemu-block-rbd"
+        end
 
-          set_boot_kernel_and_trigger_reboot('xen')
-        when "lxc"
-          package "lxc"
+        # load modules only when appropriate kernel is present
+        execute "loading kvm modules" do
+          command <<-EOF
+              grep -qw vmx /proc/cpuinfo && /sbin/modprobe kvm-intel
+              grep -qw svm /proc/cpuinfo && /sbin/modprobe kvm-amd
+              grep -q POWER /proc/cpuinfo && /sbin/modprobe kvm
+              /sbin/modprobe vhost-net
+              /sbin/modprobe nbd
+          EOF
+          only_if { %x[uname -r].include?('default') }
+        end
+      end
 
-          service "boot.cgroup" do
-            action [:enable, :start]
-          end
+    when "xen"
+      %w{kernel-xen xen xen-tools openvswitch-kmp-xen}.each do |pkg|
+        package pkg do
+          action :install
+        end
       end
 
-      # change libvirt to run qemu as user qemu
-      # make sure to only set qemu:kvm for kvm and qemu deployments, use
-      # system defaults for xen
-      if ['kvm','qemu'].include?(node[:nova][:libvirt_type])
-        libvirt_user = "qemu"
-        libvirt_group = "kvm"
-      else
-        libvirt_user = "root"
-        libvirt_group = "root"
+      service "xend" do
+        action :nothing
+        supports :status => true, :start => true, :stop => true, :restart => true
+        # restart xend only when xen kernel is already present
+        only_if { %x[uname -r].include?('xen') }
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.<br>Do not use `%x` unless the command string contains backquotes.